### PR TITLE
use dynamic_pointer_cast to detect allocator mismatch in intra process manager

### PR DESCRIPTION
**Pull request by wjwwood**
We believe this is the root cause of https://github.com/ros2/demos/issues/501

### CODE REVIEW

**Review by ivanpauno — DISMISSED** [review]
LGMT with green CI (it would be good to at least run tests for `test_rclcpp`)

**Comment by wjwwood** [comment]
@ivanpauno I'm gonna add some tests to cover this.

File: rclcpp/include/rclcpp/experimental/intra_process_manager.hpp [hunk 1]
```diff
@@ -394,9 +398,13 @@ class IntraProcessManager
       }
       auto subscription_base = subscription_it->second.subscription.lock();
       if (subscription_base) {
-        auto subscription = std::static_pointer_cast<
+        auto subscription = std::dynamic_pointer_cast<
           rclcpp::experimental::SubscriptionIntraProcess<MessageT>
```
Comment (hidmic) [inline rclcpp/include/rclcpp/experimental/intra_process_manager.hpp:402]: @wjwwood shouldn't this be
```suggestion
          rclcpp::experimental::SubscriptionIntraProcess<MessageT, Alloc, Deleter>
```
implying publisher and subscription have the same allocator and deleter? Same above (we'll need extra template parameters).

Comment (wjwwood) [reply rclcpp/include/rclcpp/experimental/intra_process_manager.hpp:402]: That might be the fix, I was gonna write a test cast first. I hadn't investigated the issue yet.

**Comment by wjwwood** [comment]
Ok, with the new test case added, they both fail:

```
$ /home/william/ros2_ws/build/rclcpp/test/rclcpp/test_intra_process_manager_with_allocators
Running main() from gmock_main.cc
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from TestIntraProcessManagerWithAllocators
[ RUN      ] TestIntraProcessManagerWithAllocators.custom_allocator
1619557665.532061 [87] test_intra: using network interface enp0s5 (udp/10.211.55.6) selected arbitrarily from: enp0s5, docker0
/home/william/ros2_ws/src/ros2/rclcpp/rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp:151: Failure
Expected: { publisher->publish(std::move(msg)); executor.spin_some(); } doesn't throw an exception.
  Actual: it throws.
[  FAILED  ] TestIntraProcessManagerWithAllocators.custom_allocator (31 ms)
[ RUN      ] TestIntraProcessManagerWithAllocators.custom_allocator_wrong
1619557665.557867 [87] test_intra: using network interface enp0s5 (udp/10.211.55.6) selected arbitrarily from: enp0s5, docker0
/home/william/ros2_ws/src/ros2/rclcpp/rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp:214: Failure
Expected: { publisher->publish(std::move(msg)); executor.spin_some(); } throws an exception of type std::runtime_error.
  Actual: it throws nothing.
[  FAILED  ] TestIntraProcessManagerWithAllocators.custom_allocator_wrong (21 ms)
[----------] 2 tests from TestIntraProcessManagerWithAllocators (52 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (53 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] TestIntraProcessManagerWithAllocators.custom_allocator
[  FAILED  ] TestIntraProcessManagerWithAllocators.custom_allocator_wrong

 2 FAILED TESTS
```

File: rclcpp/include/rclcpp/experimental/intra_process_manager.hpp (hunk 1 above)
Comment (wjwwood) [reply rclcpp/include/rclcpp/experimental/intra_process_manager.hpp:404]: I might need to take a break to work on something else, but later tonight I'll try this to see if it fixes it.

Comment (wjwwood) [reply rclcpp/include/rclcpp/experimental/intra_process_manager.hpp]: Looks like this is the solution, see: 490cca76

The tests pass now. 🎉

**Comment by wjwwood** [comment]
I may have to fix some more things with this. But I was able to build rclcpp and run the test. I haven't gotten to the allocator demo to see if that fixes the problem actually.

**Comment by wjwwood** [comment]
I tested the demo and the problem seems to be fixed now, CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14494/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9254/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12161/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14595/)

**Review by ivanpauno — APPROVED** [review]

**Comment by ivanpauno** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14497/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9257/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12165/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14598/)

File: rclcpp/include/rclcpp/experimental/intra_process_manager.hpp [hunk 2]
```diff
@@ -363,9 +367,16 @@ class IntraProcessManager
...
-          rclcpp::experimental::SubscriptionIntraProcess<MessageT>
+        auto subscription = std::dynamic_pointer_cast<
+          rclcpp::experimental::SubscriptionIntraProcess<MessageT, Alloc, Deleter>
           >(subscription_base);
+        if (nullptr == subscription) {
+          throw std::runtime_error(
+                  "failed to dynamic cast SubscriptionIntraProcessBase to "
+                  "SubscriptionIntraProcess<MessageT, Alloc, Deleter>, which "
+                  "can happen when the publisher and subscription use different "
+                  "allocator types, which is not supported");
```
Comment (hidmic) [inline rclcpp/include/rclcpp/experimental/intra_process_manager.hpp:378]: +10 to the hint

File: rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
```
Comment (hidmic) [inline rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp:1]: @wjwwood 
```suggestion
// Copyright 2015-2021 Open Source Robotics Foundation, Inc.
```
? I know half the file is old, but the other half is brand new.

File: rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+  MessageAlloc message_alloc = *alloc;
+  rclcpp::allocator::set_allocator_for_deleter(&message_deleter, &message_alloc);
+
+  auto ptr = MessageAllocTraits::allocate(message_alloc, 1);
+  MessageAllocTraits::construct(message_alloc, ptr);
+  std::unique_ptr<test_msgs::msg::Empty, MessageDeleter> msg(ptr, message_deleter);
+  EXPECT_NO_THROW(
+  {
+    publisher->publish(std::move(msg));
+    executor.spin_some();
```
Comment (hidmic) [inline rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp:154]: @wjwwood shall we put an expectation on `counter`? Like:

```
EXPECT_EQ(1u, counter);
```

**Review by hidmic — APPROVED** [review]
LGTM once CI goes green!

On a related note, using custom allocators is a bit... convoluted. We really need to improve this. I haven't seen the `polymorphic_allocator` work yet though.

**Comment by clalancette** [comment]
Confirmed that this change fixes the problem for me in the demo_nodes_cpp allocator example as well as the tlsf allocator example.

File: rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp (hunk 4 above)
Comment (hidmic) [reply rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp:154]: Also, an expectation on the message address would be nice too (to make sure the custom allocator didn't induce a copy).

Comment (wjwwood) [reply rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp]: Well, that was more complicated than I expected, but I did it 060526d :)

File: rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp (hunk 3 above)
Comment (wjwwood) [reply rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp]: I basically rewrote it b4a0aa1d

**Comment by wjwwood** [comment]
New CI (thanks @ivanpauno for fixing my mistake in the arguments before):

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14506/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9263/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12171/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14605/)

**Review by ivanpauno — APPROVED** [review]

**Comment by wjwwood** [comment]
The Rpr failure is a flaky test in the node graph tests.

File: rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+  typename SubscriptionAllocatorT,
+  typename MessageMemoryStrategyAllocatorT,
+  typename MemoryStrategyAllocatorT,
+  typename ExpectedExceptionT
+>
+void
+do_custom_allocator_test(
+  PublishedMessageAllocatorT published_message_allocator,
+  PublisherAllocatorT publisher_allocator,
+  SubscribedMessageAllocatorT /* subscribed_message_allocator */,  // intentinally unused
```
Comment (hidmic) [inline rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp:109]: @wjwwood nit: 
```suggestion
  SubscribedMessageAllocatorT /* subscribed_message_allocator */,  // intentionally unused
```
but fine to defer, no need to run CI again.

**Review by hidmic — APPROVED** [review]

**Comment by ivanpauno** [comment]
@Mergifyio backport galactic foxy

**Pull request merged**